discovery_engine_core/providers/src/models/content.rs

```diff
@@ -0,0 +1,238 @@
+// Copyright 2022 Xayn AG
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, version 3.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+use crate::{Error, NewscatcherArticle};
+use chrono::NaiveDateTime;
+use serde::{Deserialize, Serialize};
+use std::ops::Deref;
+use url::Url;
+
+/// A helper type used to ensure that, within the [`GenericArticle`] struct,
+/// we never use a URL which does not have a domain, such as `file:///foo/bar`.
+#[derive(Debug, Serialize, Deserialize, Clone)]
+pub struct UrlWithDomain(Url);
+
+impl Deref for UrlWithDomain {
+    type Target = Url;
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl UrlWithDomain {
+    pub fn new(url: Url) -> Option<Self> {
+        if url.domain().is_none() {
+            None
+        } else {
+            Some(UrlWithDomain(url))
+        }
+    }
+
+    #[allow(clippy::missing_panics_doc)]
+    pub fn domain(&self) -> &str {
+        self.0.domain().unwrap(/* constructor makes sure we have a domain */)
+    }
+
+    pub fn inner(&self) -> Url {
+        self.0.clone()
+    }
+}
+
+impl TryFrom<&str> for UrlWithDomain {
+    type Error = ();
+
+    fn try_from(value: &str) -> Result<Self, Self::Error> {
+        Ok(UrlWithDomain::new(Url::parse(value).unwrap()).unwrap())
+    }
+}
+
+/// Represents a news that is delivered by an external content API.
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct GenericArticle {
```

Comment:
I think it's worthwhile to be explicit here, so that it cannot be confused with all the other `Article` structs we have flying around, and since we're averse to calling things e.g. `NewscatcherArticle`. So I'd leave it as is, unless you 🔧 it :D

Reply:
ok i understand, then leave it as it is